tests/unit/memory_storage_client/test_dataset_client.py

```diff
@@ -138,3 +138,13 @@ async def test_iterate_items(dataset_client: DatasetClient) -> None:
     assert len(actual_items) == item_count
     assert actual_items[0]['id'] == 0
     assert actual_items[99]['id'] == 99
+
+
+async def test_reuse_dataset(dataset_client: DatasetClient, memory_storage_client: MemoryStorageClient) -> None:
+    item_count = 10
+    await dataset_client.push_items([{'id': i} for i in range(item_count)])
+
+    memory_storage_client.datasets_handled = []  # purge datasets loaded to test create_dataset_from_directory
+    datasets_client = memory_storage_client.datasets()
+    dataset_info = await datasets_client.get_or_create(name='test')
+    assert dataset_info.item_count == item_count
```

Comment:
Hi @vdusek - I see the expected assert error (extra incremented item_count) when running this test without my change
`AssertionError: assert 11 == 10`

Reply:
So what am I doing wrong that it is passing?

(I'm on the current master with no more changes than adding this new test.)

![image](https://github.com/user-attachments/assets/62a3b5e4-2113-4bbc-b805-bc4f416afb0b)

Reply:
Thanks for the screenshot, it helped to pinpoint the issue: Linux vs MacOs.
```os.scandir``` called in ```create_dataset_from_directory``` use a filesystem dependant ordering. The bug is triggered only when ```__metadata__.json``` is not the first file scanned 

OSX scandir output:
<img width="426" alt="Screenshot 2024-08-29 at 23 55 07" src="https://github.com/user-attachments/assets/f579a17c-d6c5-4c55-bd49-17d107442a8b">

Linux scandir output:
<img width="352" alt="Screenshot 2024-08-29 at 23 47 51" src="https://github.com/user-attachments/assets/e3a5707c-c7d9-4bab-9a99-f5eb49ad0b6a">

Reply:
I tried to replace all (6) occurrences of

```python
os.scandir(directory)
```

with

```python
sorted(os.scandir(directory), key=lambda entry: entry.name, reverse=True)
```

so that `__metadata__.json` is not the first file scanned.

And I am still not successful in reproducing it - the test is passing.

Reply:
try ```reverse = False``` ==> ```AssertionError: assert 20 == 10```
because 
```python
sorted('123_', reverse=True)==['_', '3', '2', '1']
```

Reply:
Oh, I don't know what I was doing... It works now, thanks - gonna merge it. Thank you!